guidance/visual/_renderer.py

```diff
@@ -0,0 +1,450 @@
+""" Renderer is responsible for displaying output.
+
+Our main focus is on jupyter notebooks and later terminal.
+"""
+# TODO(nopdive): Implementation for terminals & append-only text displays.
+# NOTE(nopdive): Testing this notebook related components is tricky. Should figure this out at some point.
+
+import asyncio
+import logging
+import weakref
+from typing import Optional, Callable
+from pydantic import BaseModel
+from asyncio import Queue
+
+from ._message import ExecutionCompletedMessage, ExecutionCompletedOutputMessage, \
+    deserialize_message, serialize_message, MetricMessage, ClientReadyAckMessage
+from ..trace import TraceHandler
+from ..visual import GuidanceMessage, TraceMessage, ResetDisplayMessage, ClientReadyMessage
+from ._trace import trace_node_to_html
+from ._async import run_async_coroutine, async_loop, async_task
+
+try:
+    from IPython.display import clear_output, display, HTML
+    from IPython import get_ipython
+
+    ipython_imported = True
+except ImportError:
+    ipython_imported = False
+
+
+try:
+    import stitch
+
+    stitch_installed = True
+except ImportError:
+    stitch_installed = False
+
+logger = logging.getLogger(__name__)
+
+
+# NOTE(nopdive): Temporary solution until we iron out all issues with widgets.
+_legacy_mode = False  # pragma: no cover
+def legacy_mode(flag: Optional[bool] = None) -> bool:  # pragma: no cover
+    """ Sets visualizations to legacy mode (how Guidance used to be).
+
+    This method call will likely deprecate when a fresher API for
+    renderers is defined.
+
+    Args:
+        flag: True if legacy is wanted.
+
+    Returns:
+        Whether legacy mode is enabled.
+    """
+    global _legacy_mode
+    if flag is not None:
+        _legacy_mode = flag
+    return _legacy_mode
+
+
+class RenderUpdate(BaseModel):
+    """ This informs which new messages are to render, along with if a new display is needed."""
+
+    messages: list = []
+    need_new_display: bool = False
+
+
+class JupyterChangeDetector:
+    """ Detects if a change has occurred on a Jupyter cell."""
+
+    def __init__(self):
+        self._prev_cell_msg_id: Optional[int] = None
+        self._prev_cell_exec_count: Optional[int] = None
+
+    def has_changed(self) -> bool:
+        """ Detects if Jupyter cell has been updated.
+
+        Returns:
+            True if cell has a higher exec count or if we're in a different cell to before.
+        """
+        if ipython_imported and get_ipython() is not None:
+            ipy = get_ipython()
+            cell_msg_id = ipy.get_parent()["msg_id"]
```

Comment:
Getting an exception here in terminal ipython:
```
AttributeError: 'TerminalInteractiveShell' object has no attribute 'get_parent'
```
Not to go down the rabbit-hole of environment detection too much, but some bare-bones handler for the terminal would be really nice (or at least an explicit exception telling me that `echo=False` must be used).

Reply:
Good catch, this is an oversight on my part, I should have tested on a terminal and promptly forgot.

I'll try to get the fix in today.

Reply:
@hudson-ai , there should be a fix for terminal now. No side-effect renders until some of the IPython shell event handling is figured out (I'll update this to the future work items list).